### PR TITLE
[SPARK-45033][SQL] Support maps by parameterized `sql()`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to allow a few more column expressions as parameters of `sql()` to construct maps/array: `CreateArray`, `MapFromArrays`, `CreateMap`, MapFromArrays` and `MapFromEntries`. Need to allow such expression because Spark SQL doesn't support constructing literals of the map type.

### Why are the changes needed?
To improve user experience with Spark SQL, and support the rest built-in type.

### Does this PR introduce _any_ user-facing change?
No. It extends the existing API.

### How was this patch tested?
By running new tests:
```
$ build/sbt "test:testOnly *ParametersSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan Could you review this PR, please.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala [hunk 1]
```diff
@@ -96,7 +96,11 @@ case class PosParameterizedQuery(child: LogicalPlan, args: Array[Expression])
  */
 object BindParameters extends Rule[LogicalPlan] with QueryErrorsBase {
   private def checkArgs(args: Iterable[(String, Expression)]): Unit = {
-    args.find(!_._2.isInstanceOf[Literal]).foreach { case (name, expr) =>
+    def isNotAllowed(expr: Expression): Boolean = expr.exists {
+      case _: Literal | _: CreateArray | _: MapFromArrays | _: CreateMap => false
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:100]: We should recursively check the children of `CreateArray`, etc.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:100]: `_.exists` checks the children recursively, doesn't it? So, we allow mix of arrays, maps and literals.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:100]: Let me add an negative test with an deep nested prohibited expression.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:100]: Ah I see! It's `TreeNode#exists`

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:100]: Can we add `CreateStruct`? It's weird that we only leave it out.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:100]: ok. How about to support `MapFromEntries` as well? And maybe `CreateNamedStruct`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:100]: `CreateNamedStruct` is the actual expression for `CreateStruct`. `MapFromEntries` also look fine to me.

**Pull request closed without merging**